File: spikes/routerless-transport/ios/WiFiAwareSpike.swift
// Milestone 4A, Phase 2 -- DRAFT, UNTESTED. See ../README.md before using this file.
//
// Minimal Wi-Fi Aware publish/subscribe/connect spike for iOS/iPadOS 26+, iPhone 12+
// only. Proves nothing about MeshTalk's mesh protocol -- it exchanges exactly one
// fixed literal string, `MESHTALK_ROUTERLESS_TEST_V1`, over a Wi-Fi Aware data path,
// per Milestone 4A Phase 2's scope.
//
// Every API used here is taken directly from Apple's own sample code and
// documentation (not guessed):
//   - "Building peer-to-peer apps": https://developer.apple.com/documentation/wifiaware/building-peer-to-peer-apps
//   - "Connecting devices for peer-to-peer Wi-Fi": https://developer.apple.com/documentation/wifiaware/connecting-paired-devices
//   - "Adopting Wi-Fi Aware": https://developer.apple.com/documentation/WiFiAware/Adopting-Wi-Fi-Aware
//
// Prerequisites this file assumes but does NOT set up (see ../README.md):
//   1. A separate Xcode target (NOT the main MeshTalk app) with the
//      `com.apple.developer.wifi-aware` entitlement, capability array `["Publish",
//      "Subscribe"]`.
//   2. That target's Info.plist declares:
//        WiFiAwareServices -> "_meshtalk-rlt._tcp" -> Publishable: {}, Subscribable: {}
//   3. Physical iPhone 12+ running iOS 26+. Apple's own sample page states you
//      "can't run this sample in Simulator". This file has never been compiled or
//      run -- treat every line as unverified until it has been.
//   4. A user-driven pairing step BEFORE any of the code below can target a specific
//      peer -- Apple's sample confirms Wi-Fi Aware connections only target *paired*
//      devices (`WAPairedDevice`), paired via `DeviceDiscoveryUI`'s `DevicePicker`
//      (subscriber side) / `DevicePairingView` (publisher side), or `AccessorySetupKit`.
//      That SwiftUI pairing flow is NOT implemented below (only sketched in comments)
//      because no full code sample for it was fetched/verified for this draft -- do
//      not guess it; read
//      https://developer.apple.com/documentation/DeviceDiscoveryUI before implementing.

import Foundation
import Network
import WiFiAware // iOS/iPadOS 26+ only -- guard all use sites with @available.
import os.log

private let logger = Logger(subsystem: "com.meshtalk.routerless-spike", category: "wifi-aware")

/// Must exactly match the `WiFiAwareServices` key declared in the spike target's
/// Info.plist (see prerequisite 2 above) -- Apple's docs warn a mismatch here is a
/// crash, not a soft failure.
private let meshTalkSpikeServiceName = "_meshtalk-rlt._tcp"

/// Fixed test payload for Milestone 4A Phase 2 -- deliberately not mesh-protocol
/// content, not encrypted, just proving a data path carries bytes at all.
private let testPayload = "MESHTALK_ROUTERLESS_TEST_V1".data(using: .utf8)!

@available(iOS 26.0, *)
extension WAPublishableService {
    static var meshTalkSpikeService: WAPublishableService {
        allServices[meshTalkSpikeServiceName]!
    }
}

@available(iOS 26.0, *)
extension WASubscribableService {
    static var meshTalkSpikeService: WASubscribableService {
        allServices[meshTalkSpikeServiceName]!
    }
}

/// One exchanged event -- intentionally trivial (Phase 2 scope: prove transport only).
@available(iOS 26.0, *)
enum SpikeEvent: Codable, Sendable {
    case testString(String)
}

@available(iOS 26.0, *)
typealias SpikeConnection = NetworkConnection<Coder<SpikeEvent, SpikeEvent, NetworkJSONCoder>>

/// Publisher role: advertises the spike service and echoes the test payload back to
/// whoever connects. Corresponds to Apple's "Create a listener to publish" sample.
///
/// TODO (blocked on hardware + pairing UI, see file header): before this can target
/// real paired devices, the app needs a `DevicePairingView(.wifiAware(.connecting(to:
/// .meshTalkSpikeService, from: .userSpecifiedDevices)))` flow (or similar -- verify
/// against https://developer.apple.com/documentation/DeviceDiscoveryUI/DevicePairingView
/// before writing it) so `WAPairedDevice.allDevices` actually contains a peer to allow
/// connections from. `.allPairedDevices` below is a placeholder for "whichever peers
/// got paired via that not-yet-written flow."
@available(iOS 26.0, *)
final class SpikePublisher {
    private var listener: NetworkListener<SpikeEvent, SpikeEvent, NetworkJSONCoder>?
    private var connections: [SpikeConnection] = []

    func start() async throws {
        let listener = try await NetworkListener(
            for: .wifiAware(.connecting(to: .meshTalkSpikeService, from: .allPairedDevices)),
            using: {
                Coder(receiving: SpikeEvent.self, sending: SpikeEvent.self, using: NetworkJSONCoder()) {
                    TLS()
                }
            }
        )
        .onStateUpdate { listener, state in
            logger.info("publisher listener state: \(String(describing: state))")
        }
        .run { [weak self] connection in
            logger.info("publisher: incoming connection")
            self?.handle(connection)
        }
        self.listener = listener
    }

    private func handle(_ connection: SpikeConnection) {
        connection.onStateUpdate { connection, state in
            logger.info("publisher connection state: \(String(describing: state))")
        }
        connections.append(connection)
        Task {
            do {
                for try await (event, _) in connection.messages {
                    if case let .testString(received) = event {
                        logger.info("publisher received: \(received, privacy: .public)")
                        // Echo it straight back so the subscriber can confirm a
                        // full round trip, not just one-way delivery.
                        try await connection.send(.testString(received))
                    }
                }
            } catch {
                logger.error("publisher connection error: \(error, privacy: .public)")
            }
        }
    }
}

/// Subscriber role: discovers the publisher, connects, sends the fixed test payload,
/// and waits for the echoed reply. Corresponds to Apple's "Create a browser to
/// subscribe" + "Make a connection" samples.
///
/// Same pairing-flow caveat as `SpikePublisher` applies -- see its doc comment.
@available(iOS 26.0, *)
final class SpikeSubscriber {
    func runOnce() async throws -> String {
        let browser = NetworkBrowser(
            for: .wifiAware(.connecting(to: .allPairedDevices, from: .meshTalkSpikeService))
        )
        .onStateUpdate { browser, state in
            logger.info("subscriber browser state: \(String(describing: state))")
        }

        let endpoint = try await browser.run { waEndpoints in
            logger.info("subscriber discovered \(waEndpoints.count) endpoint(s)")
            if let first = waEndpoints.first {
                return .finish(first)
            }
            return .continue
        }

        let connection = SpikeConnection(
            to: endpoint,
            using: {
                Coder(receiving: SpikeEvent.self, sending: SpikeEvent.self, using: NetworkJSONCoder()) {
                    TLS()
                }
            }
        )
        connection.onStateUpdate { connection, state in
            logger.info("subscriber connection state: \(String(describing: state))")
        }

        try await connection.send(.testString(String(data: testPayload, encoding: .utf8)!))

        for try await (event, _) in connection.messages {
            if case let .testString(echoed) = event {
                return echoed
            }
        }
        throw NSError(domain: "MeshTalkRoutlerlessSpike", code: 1, userInfo: [NSLocalizedDescriptionKey: "connection closed before a reply arrived"])
    }
}

File: spikes/routerless-transport/android/WifiAwareSpike.kt
// Milestone 4A, Phase 2 -- DRAFT, UNTESTED. See ../README.md before using this file.
//
// Minimal Wi-Fi Aware publish/subscribe/connect spike for Android 8.0 (API 26)+,
// devices with Wi-Fi Aware hardware actually present (chipset-dependent -- must be
// runtime-checked, see checkWifiAwareAvailable() below). Proves nothing about
// MeshTalk's mesh protocol -- it exchanges exactly one fixed literal string,
// MESHTALK_ROUTERLESS_TEST_V1, over a real Socket/ServerSocket carried across a
// Wi-Fi Aware network connection, per Milestone 4A Phase 2's scope.
//
// Every API used here is taken directly from Android's own "Wi-Fi Aware overview"
// guide (https://developer.android.com/develop/connectivity/wifi/wifi-aware),
// specifically its "Initial setup", "Obtain a session", "Publish a service",
// "Subscribe to a service", "Send a message", and "Create a connection" sections --
// not guessed. This file follows that guide's 8-step connection recipe exactly:
//   1. Publisher publishes a service; subscriber subscribes to it.
//   2. Subscriber discovers publisher, sends it a short message.
//   3. Publisher opens a ServerSocket, gets its port.
//   4. Publisher requests a Wi-Fi Aware network via ConnectivityManager +
//      WifiAwareNetworkSpecifier(discoverySession, subscriberPeerHandle, port).
//   5. Once the publisher's network request is up, it sends a message back to the
//      subscriber (the guide doesn't specify content here; this file reuses the
//      publisher's own PeerHandle-independent port number so the subscriber knows
//      networking is ready).
//   6. Subscriber receives that message, then requests its own Wi-Fi Aware network
//      (same API, no port).
//   7. Subscriber's onAvailable()/onCapabilitiesChanged() gives it the publisher's
//      IPv6 address + port via WifiAwareNetworkInfo -- opens a real Socket to the
//      ServerSocket.
//   8. Fixed test string is written/read over that socket; connections are closed.
//
// Prerequisites this file assumes but does NOT set up (see ../README.md):
//   - Manifest permissions listed in docs/routerless-transport-capability-matrix.md
//     (ACCESS_WIFI_STATE, CHANGE_WIFI_STATE, CHANGE_NETWORK_STATE, INTERNET,
//     NEARBY_WIFI_DEVICES for API 33+, ACCESS_FINE_LOCATION up to API 32) are NOT
//     currently declared in android/app/src/main/AndroidManifest.xml -- add them to
//     a throwaway module/debug screen, not the shipping manifest, until Phase 2
//     actually passes.
//   - Never compiled, never run. Treat every line as unverified until it has been
//     built and exercised on two physical devices with confirmed Wi-Fi Aware
//     hardware.

package com.meshtalk.spike.routerless

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.wifi.aware.DiscoverySessionCallback
import android.net.wifi.aware.PeerHandle
import android.net.wifi.aware.PublishConfig
import android.net.wifi.aware.PublishDiscoverySession
import android.net.wifi.aware.SubscribeConfig
import android.net.wifi.aware.SubscribeDiscoverySession
import android.net.wifi.aware.WifiAwareManager
import android.net.wifi.aware.WifiAwareNetworkInfo
import android.net.wifi.aware.WifiAwareNetworkSpecifier
import android.net.wifi.aware.WifiAwareSession
import android.util.Log
import java.net.ServerSocket
import java.net.Socket
import java.nio.charset.StandardCharsets

private const val TAG = "MeshTalkRoutlerlessSpike"

/** Must match on both publisher and subscriber. */
private const val SERVICE_NAME = "MeshTalkRoutlerlessSpikeV1"

/** Fixed test payload for Milestone 4A Phase 2 -- deliberately not mesh-protocol content. */
private const val TEST_PAYLOAD = "MESHTALK_ROUTERLESS_TEST_V1"

/** Placeholder shared secret for the Wi-Fi Aware network -- a real integration must
 * not hardcode this; not in Phase 2's scope to fix. */
private const val SPIKE_PSK = "meshtalk-routerless-spike-placeholder"

/**
 * Step 0 (not in Android's numbered recipe, but required before any of it): confirm
 * the device actually has Wi-Fi Aware hardware available right now. Per the guide,
 * `hasSystemFeature` can be true (OS supports the API) while `isAvailable()` is false
 * (Wi-Fi/Location off, or hardware busy with Wi-Fi Direct/SoftAP/tethering).
 */
fun checkWifiAwareAvailable(context: Context): Boolean {
    val hasFeature = context.packageManager.hasSystemFeature(PackageManager.FEATURE_WIFI_AWARE)
    val manager = context.getSystemService(Context.WIFI_AWARE_SERVICE) as? WifiAwareManager
    val available = manager?.isAvailable == true
    Log.i(TAG, "hasSystemFeature=$hasFeature isAvailable=$available")
    return hasFeature && available
}

/**
 * Registers for WifiAwareManager.ACTION_WIFI_AWARE_STATE_CHANGED, per the guide's
 * instruction that availability can change at any time and existing sessions should
 * be discarded when it does. Caller owns unregistering this receiver.
 */
fun registerAvailabilityReceiver(context: Context, onChanged: (Boolean) -> Unit): BroadcastReceiver {
    val manager = context.getSystemService(Context.WIFI_AWARE_SERVICE) as? WifiAwareManager
    val receiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context, intent: Intent) {
            onChanged(manager?.isAvailable == true)
        }
    }
    context.registerReceiver(receiver, IntentFilter(WifiAwareManager.ACTION_WIFI_AWARE_STATE_CHANGED))
    return receiver
}

/**
 * Publisher ("server") role. Publishes SERVICE_NAME, waits for the subscriber's
 * initial message, opens a ServerSocket, and requests a Wi-Fi Aware network scoped
 * to that specific subscriber's PeerHandle + the socket's port.
 */
class SpikePublisher(private val context: Context) {

    private var session: WifiAwareSession? = null
    private var discoverySession: PublishDiscoverySession? = null
    private var serverSocket: ServerSocket? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    fun start(onTestStringReceived: (String) -> Unit) {
        val manager = context.getSystemService(Context.WIFI_AWARE_SERVICE) as WifiAwareManager
        manager.attach(object : android.net.wifi.aware.AttachCallback() {
            override fun onAttached(awareSession: WifiAwareSession) {
                session = awareSession
                publish(awareSession, onTestStringReceived)
            }

            override fun onAttachFailed() {
                Log.e(TAG, "publisher: attach failed")
            }
        }, null)
    }

    private fun publish(awareSession: WifiAwareSession, onTestStringReceived: (String) -> Unit) {
        val config = PublishConfig.Builder()
            .setServiceName(SERVICE_NAME)
            .build()
        awareSession.publish(config, object : DiscoverySessionCallback() {
            override fun onPublishStarted(publishSession: PublishDiscoverySession) {
                Log.i(TAG, "publisher: publish started")
                discoverySession = publishSession
            }

            override fun onMessageReceived(peerHandle: PeerHandle, message: ByteArray) {
                // Subscriber's initial "hello" message (step 2). Now open the
                // ServerSocket and request the Wi-Fi Aware network (steps 3-4).
                Log.i(TAG, "publisher: message from subscriber: ${String(message, StandardCharsets.UTF_8)}")
                openServerSocketAndRequestNetwork(peerHandle, onTestStringReceived)
            }
        }, null)
    }

    private fun openServerSocketAndRequestNetwork(peerHandle: PeerHandle, onTestStringReceived: (String) -> Unit) {
        val ss = ServerSocket(0)
        serverSocket = ss
        val port = ss.localPort

        val ds = discoverySession ?: run {
            Log.e(TAG, "publisher: no discovery session, aborting")
            return
        }

        val networkSpecifier = WifiAwareNetworkSpecifier.Builder(ds, peerHandle)
            .setPskPassphrase(SPIKE_PSK)
            .setPort(port)
            .build()
        val request = NetworkRequest.Builder()
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI_AWARE)
            .setNetworkSpecifier(networkSpecifier)
            .build()

        val connMgr = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                Log.i(TAG, "publisher: network available, sending port back to subscriber")
                // Step 5: tell the subscriber networking is ready. The guide leaves
                // message content to the app; this spike just re-sends the port as
                // ASCII so the subscriber has confirmation (it already knows the
                // port from this same message in a from-scratch design, but the
                // guide's step 5 explicitly calls for a message here).
                ds.sendMessage(peerHandle, /* messageId = */ 1, port.toString().toByteArray(StandardCharsets.UTF_8))

                Thread {
                    try {
                        val client = ss.accept()
                        val received = client.getInputStream().bufferedReader().readLine()
                        Log.i(TAG, "publisher: received over socket: $received")
                        if (received != null) onTestStringReceived(received)
                        client.getOutputStream().write((received ?: "").toByteArray(StandardCharsets.UTF_8))
                        client.close()
                    } catch (e: Exception) {
                        Log.e(TAG, "publisher: socket accept/read failed", e)
                    }
                }.start()
            }
        }
        networkCallback = callback
        connMgr.requestNetwork(request, callback)
    }

    fun stop() {
        val connMgr = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        networkCallback?.let { connMgr.unregisterNetworkCallback(it) }
        serverSocket?.close()
        discoverySession?.close()
        session?.close()
    }
}

/**
 * Subscriber ("client") role. Subscribes to SERVICE_NAME, sends the discovered
 * publisher a "hello" message, waits for its port confirmation, requests its own
 * Wi-Fi Aware network, then opens a Socket to the publisher's ServerSocket and
 * writes TEST_PAYLOAD.
 */
class SpikeSubscriber(private val context: Context) {

    private var session: WifiAwareSession? = null
    private var discoverySession: SubscribeDiscoverySession? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    fun start(onEchoReceived: (String) -> Unit) {
        val manager = context.getSystemService(Context.WIFI_AWARE_SERVICE) as WifiAwareManager
        manager.attach(object : android.net.wifi.aware.AttachCallback() {
            override fun onAttached(awareSession: WifiAwareSession) {
                session = awareSession
                subscribe(awareSession, onEchoReceived)
            }

            override fun onAttachFailed() {
                Log.e(TAG, "subscriber: attach failed")
            }
        }, null)
    }

    private fun subscribe(awareSession: WifiAwareSession, onEchoReceived: (String) -> Unit) {
        val config = SubscribeConfig.Builder()
            .setServiceName(SERVICE_NAME)
            .build()
        awareSession.subscribe(config, object : DiscoverySessionCallback() {
            override fun onSubscribeStarted(subscribeSession: SubscribeDiscoverySession) {
                Log.i(TAG, "subscriber: subscribe started")
                discoverySession = subscribeSession
            }

            override fun onServiceDiscovered(peerHandle: PeerHandle, serviceSpecificInfo: ByteArray, matchFilter: List<ByteArray>) {
                Log.i(TAG, "subscriber: discovered publisher")
                // Step 2: send the publisher a short "hello" message.
                discoverySession?.sendMessage(peerHandle, /* messageId = */ 1, "hello".toByteArray(StandardCharsets.UTF_8))
            }

            override fun onMessageReceived(peerHandle: PeerHandle, message: ByteArray) {
                // Step 6: publisher's port-confirmation message arrived. Now request
                // our own Wi-Fi Aware network (no port specified, per the guide).
                Log.i(TAG, "subscriber: got publisher confirmation: ${String(message, StandardCharsets.UTF_8)}")
                requestNetworkAndSend(peerHandle, onEchoReceived)
            }
        }, null)
    }

    private fun requestNetworkAndSend(peerHandle: PeerHandle, onEchoReceived: (String) -> Unit) {
        val ds = discoverySession ?: run {
            Log.e(TAG, "subscriber: no discovery session, aborting")
            return
        }

        val networkSpecifier = WifiAwareNetworkSpecifier.Builder(ds, peerHandle)
            .setPskPassphrase(SPIKE_PSK)
            .build()
        val request = NetworkRequest.Builder()
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI_AWARE)
            .setNetworkSpecifier(networkSpecifier)
            .build()

        val connMgr = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
                // Step 7: pull the publisher's IPv6 + port from WifiAwareNetworkInfo,
                // open a real Socket, write the fixed test payload, read the echo.
                val peerAwareInfo = networkCapabilities.transportInfo as? WifiAwareNetworkInfo ?: return
                val peerIpv6 = peerAwareInfo.peerIpv6Addr ?: return
                val peerPort = peerAwareInfo.port

                Thread {
                    try {
                        val socket: Socket = network.socketFactory.createSocket(peerIpv6, peerPort)
                        socket.getOutputStream().write((TEST_PAYLOAD + "\n").toByteArray(StandardCharsets.UTF_8))
                        val echoed = socket.getInputStream().bufferedReader().readLine()
                        Log.i(TAG, "subscriber: echoed back: $echoed")
                        if (echoed != null) onEchoReceived(echoed)
                        socket.close()
                    } catch (e: Exception) {
                        Log.e(TAG, "subscriber: socket connect/write failed", e)
                    }
                }.start()
            }
        }
        networkCallback = callback
        connMgr.requestNetwork(request, callback)
    }

    fun stop() {
        val connMgr = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        networkCallback?.let { connMgr.unregisterNetworkCallback(it) }
        discoverySession?.close()
        session?.close()
    }
}
